### ISSUE TITLE
Generate sensible .gitignore when running `markbind init`

### ISSUE THREAD

**Issue by tlylt** [MarkBind/markbind#1782]
**Is your request related to a problem?**
Given that most people who use markbind would typically be using git to version control a markbind project, currently it's a tiny bit of a hassle to manually create a `.gitignore` when generating a new site using `markbind init`.

**Describe the solution you'd like**
Would be nice if a sensible .gitignore file is added. In particular, to ignore log files and maybe `.env` as well as `_site`.

**Describe alternatives you've considered**
Minimally, would be nice to provide a suitable example of a .gitignore file, perhaps as part of [GitHub .gitignore templates](https://github.com/github/gitignore)

**Additional context**
Might also want to consider generating a `package.json` file that will be able to track the `markbind-cli` version.
  - If `package.json` is added, the `.gitignore` file will need to ignore `/node_modules`

---

**TODO for PRs targeting this issue**
- [ ] generating a .gitignore when markbind init is called
- [ ] either skip or possibly override for the --convert flag (Preferably skip)
- [ ] update UG docs on this behavior

This is possibly breaking change

**Comment by kaixin-hc** [comment]
I like this idea, and think it's relatively straightforward for new sites when the repository contains only the website.

Playing devils advocate though:

- What about if markbind init is run inside a subfolder in a new project? Will this create extra hassle for users who now need to delete a .gitignore file?
- This doesn't really translate well to converted sites, right? If it's run with the -c option to convert a repo we might overwrite an existing .gitignore file?

**Comment by tlylt** [comment]
> * What about if markbind init is run inside a subfolder in a new project? Will this create extra hassle for users who now need to delete a .gitignore file?

If I'm not wrong, it is ok to have [multiple .gitignore](https://stackoverflow.com/questions/53208235/multiple-gitignore-in-subfolders).
 
> This doesn't really translate well to converted sites, right? If it's run with the -c option to convert a repo we might overwrite an existing .gitignore file?

Good point, might need to ensure that this is handled separately. Perhaps if we can detect whether there is an existing .gitignore, then we can either merge it by appending to the file or exclude our .gitignore.

**Comment by tlylt** [comment]
Proposed `.gitignore` content:
```
# Logs
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*
_markbind/logs/

# Dependency directories
node_modules/

# Production
_site/

# Env
.env
.env.local

# IDE configs
.vscode/
.idea/*
*.iml
```

**Comment by lhw-1** [comment]
Given the above discussion, it seems like the best way forward is to generate the `.gitignore` (with content as proposed by @tlylt) when `markbind init` is called, and if it is called with `--convert` with a pre-existing `.gitignore`, then we can merge the new `.gitignore` with the pre-existing one. In addition, while it is possible to have duplicate rules in a `.gitignore`, it may be even better if we can perform some form of selective merging for the `.gitignore` rules.

I am wondering if the default behavior of creating a `.gitignore` would be better, or if having an option to create one (and not create one by default) would be better. Personally I think that we should create a `.gitignore` on `init` by default while providing an option to not create one (e.g. `--no-gitignore`) for the `init` command - would this be good?

**Comment by jovyntls** [comment]
Regarding merging an existing `.gitignore` with our generated `.gitignore`: IMO if a .gitignore has been supplied by the user, we shouldn't modify it by default (the user should be asked first). Personally I wouldn't want a cli app to modify my gitignore because I would be concerned about unexpected behaviour, unless I know exactly what changes it is making in. I think duplicates are fine, but in other edge cases where I've intentionally ignored or un-ignored a file, I would get unexpected behaviour. It may be hard to account for all these edge cases, and not worth the effort especially if we could provide the user to a link to our sample .gitignore (perhaps in the UG).
**tldr:** IMO a "smart merge" is detrimental to UX because we're modifying a user's file, possibly wrongly

> Personally I think that we should create a .gitignore on init by default 

Agreed! It's a useful feature and would be needed the vast majority of the time

> while providing an option to not create one (e.g. --no-gitignore) for the init command - would this be good?

If we're going with the merging strategy, I think we should definitely offer users the option of skipping our gitignore!
If not, I feel that the --no-gitignore is not needed as the user can delete the gitignore if it's not wanted, and having more flags may add additional complexity to the setup process?

**Comment by tlylt** [comment]
Generating a .gitignore when markbind init is called seems to be a good first step. As for `--convert`, my preference would be to either skip it entirely or override it, depending on the ease of implementation. Reason being that
- if a user is not using `.gitignore`, skipping or generating does not have any obvious impact
- if a user is already using `.gitignore`, skipping means they continue to use their own one, overriding is not detrimental as well, as the git diff would show up that we have overridden the original file and they can recover from there

I would incline to have this feature shipped and then discuss the details about overriding during convert, based on real concerns and usages.

Will update the issue description to include what's discussed. Thanks @lhw-1 @jovyntls

**Comment by jovyntls** [comment]
> * if a user is already using `.gitignore`, skipping means they continue to use their own one, overriding is not detrimental as well, as the git diff would show up that we have overridden the original file and they can recover from there

The git history does provide a back-up, but I don't think users should have to recover their files using git history if we have overwritten their original file. Although they do have a way to get the file back, it may be a bit invasive to the user if we override their files without asking?
